### PR TITLE
[FLINK-8421] [DataStream] Make timer serializers reconfigurable on restore

### PR DESCRIPTION
**Pull request by tzulitai**
## What is the purpose of the change

Previously, key and namespace serializers of the `HeapInternalTimerService` were not reconfigured on restore.

In Flink 1.4.0, we removed Avro dependency, and on restore if the Avro dependency is not present, a `DummyAvroKryoSerializerClass` was registered to Kryo as a placeholder, which altered the base Kryo registrations in the `KryoSerializer`. This change required a serializer reconfiguration in order for restores to work. Effectively, this allowed the issue in the `HeapInternalTimerService` to surface.

This PR fixes this by writing also the `TypeSerializerConfigSnapshot`s of the key and namespace serializer of the `HeapInternalTimerService` into savepoints, and use them to reconfigure new serializers on restore.
Since this would change the binary format of the written timer services, this PR also uses this opportunity to properly make the format versioned.

This needs to be merged for `release-1.4` also.

More details of the change is explained below.

## Brief change log

- 1bc3cd0: A preliminary migration test that took a savepoint of a `WindowOperator` with keys that required serialization using the `KryoSerializer`. Savepoints were taken for Flink versions 1.2 and 1.3. Restoring from this savepoint in Flink 1.4 fails, and requires the following commits to pass.

- e1fd1fc: Always use the `FailureTolerantObjectInputStream` to read objects in the `InstantiationUtil.deserializeObject(...)` methods. That special stream avoids restore failures with `ClassNotFoundException` if Avro is not present, but there were leaks where during the restore process, that special input stream was not used.

- a078eed and dfa3e82: Introduced `ByteArrayPrependedInputStream` and `PostVersionedIOReadableWritable`. These are utility classes that were required to migrate the serialization format of the timer services from non-versioned to versioned.

- 4e27a4b: The main change, which adds key / namespace serializer config snapshots and use them for serializer reconfiguration on restore. This commit also makes the format versioned.

## Verifying this change

- The migration test added in 1bc3cd0 will not pass without all fixes.
- Unit tests are added for the new `ByteArrayPrependedInputStream` and `PostVersionedIOReadableWritable` classes.
- The `testSnapshotAndRestore` and `testSnapshotAndRebalancedRestore` tests in `HeapInternalTimerServiceTest` are adapted to test both versioned and previous non-versioned formats.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (**YES** / no / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by tzulitai** [comment]
@aljoscha As discussed offline, I've:
- replaced `ByteArrayPrependedInputStream` with Java's `PushbackInputStream`
- use negative values in `VERSIONED_IDENTIFIER` to be extra safe

File: flink-core/src/main/java/org/apache/flink/core/io/PostVersionedIOReadableWritable.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+		byte[] tmp = new byte[VERSIONED_IDENTIFIER.length];
+		stream.read(tmp);
+
+		if (Arrays.equals(tmp, VERSIONED_IDENTIFIER)) {
+			super.read(in);
+			read(in, true);
+		} else {
+			stream.unread(tmp);
+			read(new DataInputViewStreamWrapper(stream), false);
+		}
```
Comment (tzulitai) [inline flink-core/src/main/java/org/apache/flink/core/io/PostVersionedIOReadableWritable.java]: Not-so-nice things about this current implementation is:
1) it requires several layers of transforming back and forth between a `DataInputView` and `InputStream`, and 
2) it uses a separate `read(DataInputView, boolean)` method in order to wrap a "reset" `DataInputView` for the remaining reads.

I think the implementation would have been much more elegant if `DataInputView` has an `unread(byte[])` method, though I'm not sure how non-trivial it is to support this across all subclasses.
Maybe a food for thought for the future ..

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/InternalTimeServiceManager.java [hunk 2]
```diff
@@ -52,7 +52,7 @@
 
 	private final ProcessingTimeService processingTimeService;
 
-	private final Map<String, HeapInternalTimerService<K, N>> timerServices;
+	public final Map<String, HeapInternalTimerService<K, N>> timerServices;
```
Comment (aljoscha) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/InternalTimeServiceManager.java]: Is this only public for testing?

Comment (tzulitai) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/InternalTimeServiceManager.java]: no, this is a mistake that should be reverted.

**Comment by aljoscha** [comment]
I think this is good to go now. 👍 😃

**Comment by tzulitai** [comment]
Thanks for the reviews!
Will merge after Travis gives green.

**Pull request closed without merging**